### PR TITLE
balena-cli: fix linux binary with patchelf

### PR DESCRIPTION
**Pull request by kalebpace**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

----

Addresses the issues discussed in this PR: https://github.com/NixOS/nixpkgs/pull/194942

### CODE REVIEW

File: pkgs/tools/admin/balena-cli/default.nix [hunk 1]
```diff
@@ -21,27 +21,63 @@ let
...
 
+  version = "15.1.1";
   src = fetchzip {
     url = "https://github.com/balena-io/balena-cli/releases/download/v${version}/balena-cli-v${version}-${plat}-standalone.zip";
     inherit sha256;
   };
+in
+stdenv.mkDerivation rec {
+  pname = "balena-cli";
+  inherit version src;
```
Comment (mguentner) [inline pkgs/tools/admin/balena-cli/default.nix:33]: :+1:

File: pkgs/tools/admin/balena-cli/default.nix [hunk 2]
```diff
@@ -21,27 +21,63 @@ let
...
+          # $1 = name of variable to adjust
+          location=$(grep -obUam1 "$1" $out/bin/balena | cut -d: -f1)
+          location=$(expr $location + $var_skip)
+          value=$(dd if=$out/balena iflag=count_bytes,skip_bytes skip=$location \
+                     bs=1 count=$var_select status=none)
+          value=$(expr $shift_by + $value)
+          echo -n $value | dd of=$out/balena bs=1 seek=$location conv=notrunc
+        }
+        fix_offset PAYLOAD_POSITION
+        fix_offset PRELUDE_POSITION
```
Comment (mguentner) [inline pkgs/tools/admin/balena-cli/default.nix:77]: whatthe

**Review by mguentner — APPROVED** [review]
Looks good to me. Congrats on finding the amazing fix.

Just tested against the balena API. Seems to work.

File: pkgs/tools/admin/balena-cli/default.nix [hunk 3]
```diff
@@ -21,27 +21,63 @@ let
...
     url = "https://github.com/balena-io/balena-cli/releases/download/v${version}/balena-cli-v${version}-${plat}-standalone.zip";
     inherit sha256;
   };
+in
+stdenv.mkDerivation rec {
+  pname = "balena-cli";
+  inherit version src;
 
   installPhase = ''
-    runHook preInstall
```
Comment (viraptor) [inline pkgs/tools/admin/balena-cli/default.nix:35]: Why are these removed?

**Review by viraptor — COMMENTED** [review]
A question, but like good otherwise

File: pkgs/tools/admin/balena-cli/default.nix (hunk 2 above)
Comment (superherointj) [reply pkgs/tools/admin/balena-cli/default.nix:77]: Congrats for the fix.

**Comment by superherointj** [comment]
Would be nice to add a version test because as version was failing before (I tried to test it). So if something changes, that is a simple way to have some testing going on to warn.

**Comment by superherointj** [comment]
> > Would be nice to add a version test because as version was failing before (I tried to test it). So if something changes, that is a simple way to have some testing going on to warn.
> 
> Absolutely!
> 
> Would it be best to include this as a `passthru.tests` using https://ryantm.github.io/nixpkgs/contributing/coding-conventions/#ssec-inline-package-tests-writing https://ryantm.github.io/nixpkgs/builders/testers/#chap-testers
> 
> Or should it be included as a `postInstallCheck`?
> 
> I'm inclined towards the former, but while I have been experimenting, it seems that `passthru` does not prevent the package from building if the version fails

Version test is added to `passthru.tests`. It won't affect builds. (for good reasons) But nixpkgs-review and hydra runs it.

https://nixos.org/manual/nixpkgs/stable/#tester-testVersion

File: pkgs/tools/admin/balena-cli/default.nix (hunk 3 above)
Comment (superherointj) [reply pkgs/tools/admin/balena-cli/default.nix:35]: runHook aren't added to post/pre phases. But for the main phase. (installPhase, checkPhase, buildPhase, etc.) they are added, which triggers pre/post phases. Because you are replacing the installPhase, the hooks have to be manually added.

I think it is possible to improve on this. By not relying on main phase to run hooks. But then, something else would have to be used to mess with hooks. But that would be a thread on it's own.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 222139` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>balena-cli</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Comment by superherointj** [comment]
Now we wait for CI... 🪑 ☕

File: pkgs/tools/admin/balena-cli/default.nix (hunk 3 above)
Comment (kalebpace) [reply pkgs/tools/admin/balena-cli/default.nix:35]: Ah, ok makes sense. Really appreciate the clarification!

**Comment by superherointj** [comment]
⌚ Two hours later:
image

File: pkgs/tools/admin/balena-cli/default.nix [hunk 4]
```diff
@@ -21,27 +21,76 @@ let
...
+    '';
+    inherit version;
+  };
+
+  # https://github.com/NixOS/nixpkgs/pull/48193/files#diff-b65952dbe5271c002fbc941b01c3586bf5050ad0e6aa6b2fcc74357680e103ea
+  preFixup =
+    if stdenv.isLinux then
+      let
+        libPath = lib.makeLibraryPath [ stdenv.cc.cc ];
+      in
```
Comment (AndersonTorres) [inline pkgs/tools/admin/balena-cli/default.nix:59]:
```suggestion
  preFixup =
    let
        libPath = lib.makeLibraryPath [ stdenv.cc.cc ];
      in lib.optionalString stdenv.isLinux
```

File: pkgs/tools/admin/balena-cli/default.nix [hunk 5]
```diff
@@ -21,27 +21,76 @@ let
...
+          location=$(grep -obUam1 "$1" $out/bin/balena | cut -d: -f1)
+          location=$(expr $location + $var_skip)
+          value=$(dd if=$out/balena iflag=count_bytes,skip_bytes skip=$location \
+                     bs=1 count=$var_select status=none)
+          value=$(expr $shift_by + $value)
+          echo -n $value | dd of=$out/balena bs=1 seek=$location conv=notrunc
+        }
+        fix_offset PAYLOAD_POSITION
+        fix_offset PRELUDE_POSITION
+      '' else '''';
```
Comment (AndersonTorres) [inline pkgs/tools/admin/balena-cli/default.nix:91]:
```suggestion
      '';
```
